llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -8286,6 +8334,10 @@ SDValue DAGCombiner::visitOR(SDNode *N) {
     if (SDValue R = foldLogicTreeOfShifts(N, N0, N1, DAG))
       return R;
 
+  if (!TLI.hasAndNot(SDValue(N, 0)) && VT.isScalarInteger() && VT != MVT::i1)
```

Comment:
Check the simpler conditions first. could sink the hasAndNot after the pattern checks inside foldMaskedMerge